File: homework4/Satyam_Mishra/Q9_AdoptAPet.java
// Technique - QUEUE
// TIME COMPLEXITY - O(Klog K + E)
// SPACE COMPLEXITY - O(K + E)

package homework4.Satyam_Mishra;

import java.util.ArrayDeque;
import java.util.Arrays;
import java.util.Queue;

public class Q9_AdoptAPet {

    public static void main(String[] args) {
        Queue<String> dogs = new ArrayDeque<>();
        Queue<String> cats = new ArrayDeque<>();

        String[][] initialPets = {
            {"Lola", "dog", "1"},
            {"Sadie", "dog", "4"},
            {"Woof", "cat", "7"},
            {"Chirpy", "dog", "2"}
        };

        Arrays.sort(initialPets, (a, b) -> Integer.parseInt(b[2]) - Integer.parseInt(a[2]));

        for (String[] pet : initialPets) {
            if (pet[1].equals("dog")) dogs.add(pet[0]);
            else if (pet[1].equals("cat")) cats.add(pet[0]);
        }

        String[][] events = {
            {"Bob", "person", "dog"},
            {"Floofy", "cat"},
            {"Sally", "person", "cat"},
            {"Ji", "person", "cat"},
            {"Ali", "person", "cat"}
        };

        for (String[] event : events) {
            if (event.length == 3) {
                String species = event[2];
                if(species.equals("dog")) {
                    if(!dogs.isEmpty()) {
                        System.out.println(dogs.poll() + ", dog");
                    } else {
                        System.out.println(cats.poll() + ", cat");
                    }
                } else {
                    if(!cats.isEmpty()) {
                        System.out.println(cats.poll() + ", cat");
                    } else {
                        System.out.println(dogs.poll() + ", dog");
                    }
                }
                
            } else if (event.length == 2) {
                String name = event[0];
                String species = event[1];
                
                if(species.equals("dog")) {
                    dogs.add(name);
                } else {
                    cats.add(name);
                }
            }
        }
    }
}

// time taken - 20m

File: homework4/Aditya_Prasad/q9_adoptAPet.cpp
// Technique : Maintain two heaps
// T.C : O(n*logn)
// M.C : O(n)


#include <bits/stdc++.h>
using namespace std;

void runShelter(vector<vector<string>> &initialPets, vector<vector<string>> &events) {
    priority_queue<pair<pair<int, string>, string>, vector<pair<pair<int, string>, string>>> cats, dogs;

    for (int i = 0; i < initialPets.size(); i++) {
        string name = initialPets[i][0];
        string species = initialPets[i][1];
        int time = stoi(initialPets[i][2]);

        pair<pair<int, string>, string> pet = {{time, name}, species};

        if (species == "cat") {
            cats.push(pet);
        } else {
            dogs.push(pet);
        }
    }

    for (int i = 0; i < events.size(); i++) {
        if (events[i][1] == "person") {
            string desiredPet = events[i][2];
            if (desiredPet == "cat") {
                if (!cats.empty()) {
                    auto adopted = cats.top();
                    cats.pop();
                    cout << adopted.first.second << " " << adopted.second << endl;
                }
                else if (!dogs.empty()) {
                    auto adopted = dogs.top();
                    dogs.pop();
                    cout << adopted.first.second << " " << adopted.second << endl;
                }
            }
            else if (desiredPet == "dog") {
                if (!dogs.empty()) {
                    auto adopted = dogs.top();
                    dogs.pop();
                    cout << adopted.first.second << " " << adopted.second << endl;
                }
                else if (!cats.empty()) {
                    auto adopted = cats.top();
                    cats.pop();
                    cout << adopted.first.second << " " << adopted.second << endl;
                }
            }
        }
        else if (events[i][1] == "animal") {
            string name = events[i][0];
            string species = events[i][2];
            int time = stoi(events[i][3]);

            pair<pair<int, string>, string> pet = {{time, name}, species};

            if (species == "cat") {
                cats.push(pet);
            } else {
                dogs.push(pet);
            }

            cout << endl;
        }
    }
}

int main() {

    vector<vector<string>> initialPets = {
        {"Lola", "dog", "1"},
        {"Sadie", "dog", "4"},
        {"Woof", "cat", "7"},
        {"Chirpy", "dog", "2"}
    };

    vector<vector<string>> events = {
        {"Bob", "person", "dog"},
        {"Floofy", "animal", "cat", "0"},
        {"Sally", "person", "cat"},
        {"Ji", "person", "cat"},
        {"Ali", "person", "cat"}
    };

    runShelter(initialPets, events);

    return 0;
}

// Time Taken : 30 mins